File: media/editable_stuff/sdl_exercise_1.hs
import Data.Word
import Data.Bits
import Foreign
import Control.Monad.State

import qualified Graphics.UI.SDL as SDL
import qualified Graphics.UI.SDL.Image as SDL_Image
import qualified System.Environment as SysEnv

-- Haaaaaack because I don't want to go to the extra bother of being able to maniuplate other kinds of images.
-- My goal, after all, isn't to be a general purpose image thingy. I can always convert stuff separately.
rShift = 0
gShift = 8
bShift = 16
aShift = 24
rMask = 255 `shift` rShift
gMask = 255 `shift` gShift
bMask = 255 `shift` bShift
aMask = 255 `shift` aShift

setR :: Word8 -> SDL.Pixel -> SDL.Pixel
setR new (SDL.Pixel pdata) = SDL.Pixel $ (pdata .&. complement rMask) + ((fromIntegral new) :: Word32) `shift` rShift
setG :: Word8 -> SDL.Pixel -> SDL.Pixel
setG new (SDL.Pixel pdata) = SDL.Pixel $ (pdata .&. complement gMask) + ((fromIntegral new) :: Word32) `shift` gShift
setB :: Word8 -> SDL.Pixel -> SDL.Pixel
setB new (SDL.Pixel pdata) = SDL.Pixel $ (pdata .&. complement bMask) + ((fromIntegral new) :: Word32) `shift` bShift
setA :: Word8 -> SDL.Pixel -> SDL.Pixel
setA new (SDL.Pixel pdata) = SDL.Pixel $ (pdata .&. complement aMask) + ((fromIntegral new) :: Word32) `shift` aShift
getR :: SDL.Pixel -> Word8
getR (SDL.Pixel pdata) = fromIntegral $ (pdata .&. rMask) `shift` (negate rShift)
getG :: SDL.Pixel -> Word8
getG (SDL.Pixel pdata) = fromIntegral $ (pdata .&. gMask) `shift` (negate gShift)
getB :: SDL.Pixel -> Word8
getB (SDL.Pixel pdata) = fromIntegral $ (pdata .&. bMask) `shift` (negate bShift)
getA :: SDL.Pixel -> Word8
getA (SDL.Pixel pdata) = fromIntegral $ (pdata .&. aMask) `shift` (negate aShift)

getPixel :: Int -> Int -> SDL.Surface -> IO SDL.Pixel
getPixel x y surf = do
  pixels <- castPtr `liftM` SDL.surfaceGetPixels surf
  SDL.Pixel `liftM` peekElemOff pixels (x + (y * SDL.surfaceGetWidth surf))

setPixel :: Int -> Int -> SDL.Pixel -> SDL.Surface -> IO ()
setPixel x y (SDL.Pixel pdata) surf = do
  pixels <- castPtr `liftM` SDL.surfaceGetPixels surf
  pokeElemOff pixels (x + (y * SDL.surfaceGetWidth surf)) pdata

main :: IO ()
main = do
    args <- SysEnv.getArgs
    case args of
      []     -> print "Please give an image path name"
      name:_ -> hackAround name

hackAround :: String -> IO ()
hackAround filename = SDL.withInit [SDL.InitEverything] $ do
    SDL.setVideoMode 640 640 32 []
    SDL.setCaption "wait why am I setting a caption" "this is pointless"
    screen <- SDL.getVideoSurface
    image <- SDL_Image.load filename
    sequence [do
      pixel <- getPixel x y image
      setPixel x y (setR 255 . setA 128 . setB (getG pixel) $ pixel) image |
      x <- [0..(SDL.surfaceGetWidth image - 1)],
      y <- [0..(SDL.surfaceGetHeight image - 1)]]
    SDL.saveBMP image "output.bmp"
    let
      loop 30 = return ()
      loop i = do
        SDL.blitSurface image (Just $ SDL.Rect 0 0 (SDL.surfaceGetWidth image) (SDL.surfaceGetHeight image)) screen (Just $ SDL.Rect (10+(10*i)) (10+(5*i)) 0 0)
        SDL.flip screen
        SDL.delay 500
        loop (i + 1)
      in loop 0

File: media/editable_stuff/sdl_exercise_1.cpp
#include <cstdlib>
#include "SDL.h"
#include "SDL_image.h"
#include <string>
#include <inttypes.h>
#include <iostream>

class Image {
public:
  Image(SDL_Surface *internal):internal(internal){  }
  int getr(int x, int y)const;
  int getg(int x, int y)const;
  int getb(int x, int y)const;
  int geta(int x, int y)const;
  void setr(int x, int y, uint8_t val);
  void setg(int x, int y, uint8_t val);
  void setb(int x, int y, uint8_t val);
  void seta(int x, int y, uint8_t val);
private:
  uint32_t& getpixel(int x, int y);
  uint32_t const& getpixel(int x, int y)const;
  SDL_PixelFormat const* fmt()const { return internal->format; }
  SDL_Surface *internal;
};

uint32_t& Image::getpixel(int x, int y) {
  return ((uint32_t*)(internal->pixels))[x + y*internal->w];
}
uint32_t const& Image::getpixel(int x, int y)const {
  return ((uint32_t*)(internal->pixels))[x + y*internal->w];
}

int Image::getr(int x, int y)const {
  return (((getpixel(x, y) & fmt()->Rmask) >> fmt()->Rshift) << fmt()->Rloss);
}
int Image::getg(int x, int y)const {
  return (((getpixel(x, y) & fmt()->Gmask) >> fmt()->Gshift) << fmt()->Gloss);
}
int Image::getb(int x, int y)const {
  return (((getpixel(x, y) & fmt()->Bmask) >> fmt()->Bshift) << fmt()->Bloss);
}
int Image::geta(int x, int y)const {
  return (((getpixel(x, y) & fmt()->Amask) >> fmt()->Ashift) << fmt()->Aloss);
}
void Image::setr(int x, int y, uint8_t val) {
  getpixel(x, y) = (getpixel(x, y) & ~(fmt()->Rmask)) + ((val >> fmt()->Rloss) << fmt()->Rshift);
}
void Image::setg(int x, int y, uint8_t val) {
  getpixel(x, y) = (getpixel(x, y) & ~(fmt()->Gmask)) + ((val >> fmt()->Gloss) << fmt()->Gshift);
}
void Image::setb(int x, int y, uint8_t val) {
  getpixel(x, y) = (getpixel(x, y) & ~(fmt()->Bmask)) + ((val >> fmt()->Bloss) << fmt()->Bshift);
}
void Image::seta(int x, int y, uint8_t val) {
  getpixel(x, y) = (getpixel(x, y) & ~(fmt()->Amask)) + ((val >> fmt()->Aloss) << fmt()->Ashift);
}

 
int main(int argc, char **argv) {
  if (argc < 2) {
    std::cerr << "Please give an image path name\n";
    return 1;
  }

  if (SDL_Init(SDL_INIT_VIDEO) != 0) {
    std::cerr << "SDL_Init(SDL_INIT_VIDEO) failed: " << SDL_GetError() << "\n";
    return 1;
  }
  atexit(SDL_Quit);
  
  SDL_Surface *screen = SDL_SetVideoMode(640, 640, 32, 0x0);
  if (screen == NULL) {
    std::cerr << "SDL_SetVideoMode failed: " << SDL_GetError() << "\n";
    return 1;
  }

  SDL_Surface *img = IMG_Load(argv[1]);
  if (img == NULL) {
    std::cerr << "Error: file could not be opened: " << IMG_GetError() << "\n";
    return 1;
  }

  SDL_LockSurface(img);
  Image superimg(img);
  for (int x = 0; x < img->w; ++x) {
    for (int y = 0; y < img->w; ++y) {
      superimg.setr(x, y, 255);
      superimg.setb(x, y, superimg.getg(x, y));
      superimg.seta(x, y, 128);
    }
  }
  SDL_UnlockSurface(img);

  SDL_SaveBMP(img, "output.bmp");
  for (int i = 0; i < 30; ++i) {
    SDL_Rect src, dst;
 
    src.x = 0;  src.y = 0;  src.w = img->w;  src.h = img->h;
    dst.x = 10+10*i;  dst.y = 10+5*i;  dst.w = 0;  dst.h = 0;
    SDL_BlitSurface(img, &src, screen, &dst);
    SDL_Flip(screen);
    SDL_Delay(500);
  }
 
  SDL_FreeSurface(img);
  return 0;
}

